File: 01 - Defeat the Algorithm Interview/E627.LongestPalindrome.py
class Solution:
    """
    @param s: a string which consists of lowercase or uppercase letters
    @return: the length of the longest palindromes that can be built
    """
    def longestPalindrome(self, s):
        # write your code here

        # method 1
        hash = {}

        for char in s:
            if char in hash:
                del hash[char]
            else:
                hash[char] = 1

        hash_size = len(hash)
        if hash_size > 0:
            hash_size -= 1

        return len(s) - hash_size

        # # method 2
        # hash, l, r = set(), 0, len(s) - 1
        # while l <= r:
        #     if s[l] in hash:
        #         hash.remove(s[l])
        #     else:
        #         hash.add(s[l])

        #     if r != l:
        #         if s[r] in hash:
        #             hash.remove(s[r])
        #         else:
        #             hash.add(s[r])

        #     l += 1
        #     r -= 1

        # hash_size = len(hash)
        # if hash_size:
        #     hash_size -= 1

        # return len(s) - hash_size

        # # method 3
        # hash = set()
        #
        # for char in s:
        #     if char in hash:
        #         hash.remove(char)
        #     else:
        #         hash.add(char)
        #
        # hash_size = len(hash)
        # if hash_size:
        #     hash_size -= 1
        #
        # return len(s) - hash_size

File: 01 - Defeat the Algorithm Interview/E627.LongestPalindrome.java
public class Solution {
    /**
     * @param s: a string which consists of lowercase or uppercase letters
     * @return: the length of the longest palindromes that can be built
     */
    public int longestPalindrome(String s) {
        // write your code here
        // // method 1:
        // Set<Character> hashset = new HashSet<>();

        // for (char ch: s.toCharArray()) {
        //     if (hashset.contains(ch)) {
        //         hashset.remove(ch);
        //     } else {
        //         hashset.add(ch);
        //     }
        // }

        // int size = s.length();
        // if (hashset.size() > 0) size -= hashset.size() - 1;

        // return size;

        // method 2:
        int[] set = new int[128];

        for (char ch: s.toCharArray()) {
            set[ch]++;
        }

        int count = 0;
        for (int c: set) {
            count += c / 2 * 2;
            if (count % 2 == 0 && c % 2 == 1) {
                count++;
            }
        }

        return count;
    }
}
